src/Component/QueryBuilder.tsx

```diff
@@ -0,0 +1,298 @@
+import { InlineField, Input, Select } from '@grafana/ui';
+import { find, includes, isArray, isString } from 'lodash';
+import { GenericOptions, GrafanaQuery, MetricConfig, MetricPayloadConfig } from '../types';
+import { EditorField, EditorFieldGroup } from '@grafana/experimental';
+
+interface LastQuery {
+  payload: string | { [key: string]: any };
+  metric: string;
+}
+
+type Props = QueryEditorProps<DataSource, GrafanaQuery, GenericOptions>;
+
+export const QueryBuilder: ComponentType<Props> = (props) => {
+  const { datasource, onChange, onRunQuery, query } = props;
+  const [metric, setMetric] = React.useState<SelectableValue<string | number>>();
+  const [payload, setPayload] = React.useState(query.payload ?? '');
+
+  const [lastQuery, setLastQuery] = React.useState<LastQuery | null>(null);
+  const [payloadConfig, setPayloadConfig] = React.useState<MetricPayloadConfig[]>([]);
+  const [metricOptions, setMetricOptions] = React.useState<MetricConfig[]>([]);
+  const [isMetricOptionsLoading, setIsMetricOptionsLoading] = React.useState<boolean>(false);
+
+  const loadMetrics = React.useCallback(() => {
+    return datasource.listMetrics(metric?.value ?? query.target ?? '', payload).then(
+      (metrics) => {
+        const foundMetric = find(metrics, (metric) => metric.value === query.target);
+        setMetric(foundMetric === undefined ? { label: '', value: '' } : foundMetric);
+        setPayloadConfig(foundMetric?.payloads ?? []);
+        return metrics;
+      },
+      (response) => {
+        setMetric({ label: '', value: '' });
+        setMetricOptions([]);
+        throw new Error(response.statusText);
+      }
+    );
+  }, [datasource, payload, metric?.value]);
+
+  const loadMetricOptions = React.useCallback(() => {
+    setIsMetricOptionsLoading(true);
+    loadMetrics()
+      .then((metrics) => {
+        setMetricOptions(metrics);
+      })
+      .finally(() => {
+        setIsMetricOptionsLoading(false);
+      });
+  }, [loadMetrics, setMetricOptions, setIsMetricOptionsLoading]);
+
+  // Initializing metric options
+  React.useEffect(() => loadMetricOptions(), []);
+
+  React.useEffect(() => {
+    if (metric?.value === undefined || metric?.value === '') {
+      return;
+    }
+
+    if (lastQuery !== null && metric?.value === lastQuery.metric && payload === lastQuery.payload) {
+      return;
+    }
+
+    setLastQuery({ payload, metric: metric.value.toString() });
+
+    onChange({ ...query, payload, target: metric.value.toString() });
+
+    onRunQuery();
+  }, [payload, metric]);
+
+  const changePayload = (
+    name: string,
+    reloadMetric: boolean,
+    v: SelectableValue<string | number> | Array<SelectableValue<string | number>>
+  ) => {
+    setPayload((ori) => {
+      let newPayload: { [key: string]: any } = {};
+      try {
+        newPayload = isString(ori) ? JSON.parse(ori) : { ...ori };
+      } catch (error) {
+        console.log(error);
```

Comment:
Can we early exit here? and use console.error instead of log, I guess

Reply:
There is no need to leave in advance. If `JSON.parse` throws an exception, the payload will be reset to an empty object ({})

Reply:
even better